src/box/relay.cc

```diff
@@ -1211,8 +1215,11 @@ static void
 relay_send_row(struct xstream *stream, struct xrow_header *packet)
 {
 	struct relay *relay = container_of(stream, struct relay, stream);
+	assert(relay->cord.id != 0);
+	assert(relay->cord.fiber->f == relay_subscribe_f ||
+	       relay->cord.fiber->f == relay_final_join_f);
 	/* Do not send heartbeats during a final join. */
-	if (relay->replica != NULL)
+	if (relay->cord.fiber->f == relay_subscribe_f)
```

Comment:
There're other ways to check whether the function is executed during join or subscribe, but this one seems to be the best from the side of code readability.

1. We can set different cord's names for join and subscribe (now it's the same on, which is created from fd). This one is not selected, as comparing strings is very costly.
2. We can check `replica->is_relay_healthy` as the status is not updated during final join, so it'll equal to false. It so no-obvious, don't really like that solution

Reply:
I like the current approach.
Just let's rewrite it a bit:
```diff
diff --git a/src/box/relay.cc b/src/box/relay.cc
index d841db0b6..9f83bfecd 100644
--- a/src/box/relay.cc
+++ b/src/box/relay.cc
@@ -1215,11 +1215,12 @@ static void
 relay_send_row(struct xstream *stream, struct xrow_header *packet)
 {
        struct relay *relay = container_of(stream, struct relay, stream);
-       assert(relay->cord.id != 0);
-       assert(relay->cord.fiber->f == relay_subscribe_f ||
-              relay->cord.fiber->f == relay_final_join_f);
+       assert(cord() == &relay->cord);
+       assert(fiber()->f == relay_subscribe_f ||
+              fiber()->f == relay_final_join_f);
+       bool is_subscribe = fiber()->f == relay_subscribe_f;
        /* Do not send heartbeats during a final join. */
-       if (relay->cord.fiber->f == relay_subscribe_f)
+       if (is_subscribe)
                relay_send_heartbeat_on_timeout(relay);
        if (packet->group_id == GROUP_LOCAL) {
                /*
@@ -1254,8 +1255,7 @@ relay_send_row(struct xstream *stream, struct xrow_header *packet)
         * it). In the latter case packet's LSN is less than or equal to
         * local master's LSN at the moment it received 'SUBSCRIBE' request.
         */
-       if (relay->cord.fiber->f == relay_final_join_f ||
-           packet->replica_id != relay->replica->id ||
+       if (!is_subscribe || packet->replica_id != relay->replica->id ||
            packet->lsn <= vclock_get(&relay->local_vclock_at_subscribe,
                                      packet->replica_id)) {
                struct errinj *inj = errinj(ERRINJ_RELAY_BREAK_LSN,
```